### PR TITLE
fix: helm parameters show effective values from inline values

### LINKED ISSUE #9213: ArgoCD UI shows incorrect values when using ArgoCD + Helm for deployment
**Describe the bug**

When deploying applications using **Helm + ArgoCD** and passing Helm values from ArgoCD manifest file to override the default helm values, the **PARAMETERS** section in the **PARAMETERS** tab still shows the default Helm chart values. 

The Application does get created with helm values that are passed from the ArgoCD manifest file, so we are good there.
But, would like to understand if this is the default behavior of ArgoCD when put together with Helm or it should show the values that are sent from ArgoCD manifest file.

**To Reproduce**
Override default Helm values from ArgoCD manifest files using:
```
source:
    helm:
      values: |
         values...
```

**Expected behavior**
The **PARAMETERS** section in the **PARAMETERS** tab should show values that were passed from the ARgoCD manifest files. If not, the UI should indicate saying that it is default helm values.

**Screenshots**
image

**Version**

```shell
v2.3.3+
```

### PR DESCRIPTION
**Pull request by nijdarshan**
## Summary
Fixes #9213

PARAMETERS section in UI showed chart defaults instead of actual values when using inline `helm.values`. Now merges user's inline values and parameter overrides into displayed parameters.

## Changes
- Added `flattenValues()` helper to convert nested YAML to dot-notation
- Modified `populateHelmAppDetails()` to merge inline values and parameters into the params map before display

## Test plan
- [x] Local lint passes
- [x] Unit tests pass
- [x] Manual test: created app with `values: replicaCount: 10` - PARAMETERS now shows `replicaCount: 10` instead of chart default `1`

Signed-off-by: Darshan Nij <nijdarshan@gmail.com>

### CODE REVIEW

**Comment by nijdarshan** [comment]
Current version:

Screenshot 2026-01-11 at 07 41 57
Screenshot 2026-01-11 at 07 42 21

---

Fixed:

Screenshot 2026-01-11 at 07 42 57

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -2281,6 +2306,22 @@ func populateHelmAppDetails(res *apiclient.RepoAppDetailsResponse, appPath strin
...
 		return err
 	}
+
+	// Merge user's inline values into params (overrides chart defaults)
+	if q.Source.Helm != nil && !q.Source.Helm.ValuesIsEmpty() {
+		var userVals map[string]any
+		if err := yaml.Unmarshal(q.Source.Helm.ValuesYAML(), &userVals); err == nil {
+			flattenValues(userVals, params, "")
+		}
+	}
```
Comment (Mangaal) [inline reposerver/repository/repository.go:2656]: Can we add a warning/log message when spec.source.helm.values is not a valid Helm values object?

**Review by Mangaal — COMMENTED** [review]
I have verified the changes locally and it is working as expected.
Could you please add some tests to cover this new logic? I have also left a small suggestion, please feel free to resolve it.

File: reposerver/repository/repository.go (hunk 1 above)
Comment (nijdarshan) [reply reposerver/repository/repository.go:2656]: Added warning log when helm values parsing fails - will now log `failed to parse helm values: <error>` if the inline values YAML is malformed.

**Comment by nijdarshan** [comment]
> I have verified the changes locally and it is working as expected. Could you please add some tests to cover this new logic? I have also left a small suggestion, please feel free to resolve it.

Added 3 unit tests:
- `TestFlattenValues` - tests the helper function with simple, nested, deeply nested, array values, and empty map cases
- `Test_populateHelmAppDetails_WithInlineValues` - verifies inline values are merged into parameters
- `Test_populateHelmAppDetails_WithParameterOverrides` - verifies --set parameters override inline values (precedence)

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -2281,6 +2306,24 @@ func populateHelmAppDetails(res *apiclient.RepoAppDetailsResponse, appPath strin
...
+
+	// Merge user's inline values into params (overrides chart defaults)
+	if q.Source.Helm != nil && !q.Source.Helm.ValuesIsEmpty() {
+		var userVals map[string]any
+		if err := yaml.Unmarshal(q.Source.Helm.ValuesYAML(), &userVals); err != nil {
+			log.Warnf("failed to parse helm values: %v", err)
+		} else {
+			flattenValues(userVals, params, "")
+		}
+	}
```
Comment (ppapapetrou76) [inline reposerver/repository/repository.go:2656]: The code unmarshals inline values into a `map[string]any` then calls `flattenValues(userVals, params, "")`
If the YAML is not a map at the document root (e.g., a top-level array), unmarshalling into map[string]any will yield nil (no error), and flattenValues will be called with nil, which will panic.

Also nested maps may sometimes have non-string key types (map[interface{}]interface{}) depending on the yaml parsing library

I recommend using the `any` type and change the code to something like this 

```suggestion
	if q.Source.Helm != nil && !q.Source.Helm.ValuesIsEmpty() {
    	var raw any
    	if err := yaml.Unmarshal(q.Source.Helm.ValuesYAML(), &raw); err != nil {
        	log.Warnf("failed to parse helm values: %v", err)
    	} else {
        	// flatten supports arbitrary root types and map[interface{}] keys
        	flattenAnyValues(raw, params, "")
    	}
	}
```

Comment (nijdarshan) [reply reposerver/repository/repository.go:2656]: aaah, you are right! lemme fix this.

Comment (nijdarshan) [reply reposerver/repository/repository.go:2656]: Fixed in latest commit. Now using `any` type with type assertion:

```go
var raw any
if err := yaml.Unmarshal(..., &raw); err != nil {
    log.Warnf("failed to parse helm values: %v", err)
} else if userVals, ok := raw.(map[string]any); ok {
    flattenValues(userVals, params, "")
} else if raw != nil {
    log.Warnf("helm values is not a map, got %T", raw)
}
```

Added edge case tests in `Test_populateHelmAppDetails_WithInvalidValues`:
- `top-level array should not panic`
- `scalar value should not panic`
- `malformed yaml should not panic`

**Review by ppapapetrou76 — COMMENTED** [review]
LGTM 🏅  - thanks for addressing my comments

**Comment by alexberry** [comment]
Is this likely to get merged? We would love this as we are having to effectively rewrite our values yaml files to dot notation to work around this bug.

**Review by ppapapetrou76 — APPROVED** [review]
LGTM 🏅 - thanks for addressing my comments - please resolve the conflicts

**Comment by alexberry** [comment]
@ppapapetrou76 I am minded to create my own fork and raise a new PR referencing this in order to get this merged, I will try and maintain attribution to @nijdarshan when resolving conflicts and testing. Would you be alright with that approach?

**Comment by nijdarshan** [comment]
> @ppapapetrou76 I am minded to create my own fork and raise a new PR referencing this in order to get this merged, I will try and maintain attribution to @nijdarshan when resolving conflicts and testing. Would you be alright with that approach?

Hey, let me look at resolving conflicts and closing this by EOD. Thank you!

**Comment by nijdarshan** [comment]
@ppapapetrou76 can you check this please, thanks!

FYI - @alexberry

**Comment by ppapapetrou76** [comment]
I enabled auto-merge but you need someone from the @argoproj/argocd-approvers group to give you a 👍🏾

**Comment by alexberry** [comment]
Thank you all for your work in getting this merged, though it appears to be very difficult to achieve :(

**Comment by alexberry** [comment]
For the benefit of engineers at argo project, when deciding priorities, this is the horrible local workaround that I will be able to get rid of once this is merged - a yaml-to-dot-notation python script that runs on every terraform run. @crenshaw-dev (as the most active member of argo-proj I could see on this repository) could we maybe have some eyes on it please?

```python
#!/usr/bin/env python3
"""
Flatten nested YAML/JSON structure to dot notation.
Reads JSON query from stdin (Terraform external data source format).
Outputs flattened JSON to stdout.
"""

import json
import sys


def escape_helm_value(v):
    """
    Escape a value for use in Helm's --set parameter.

    Based on Helm's escaping requirements for --set command line arguments.
    Only backslash and braces need escaping in parameter values.

    Args:
        v: Value to escape (should not be a list or dict)

    Returns:
        Properly escaped string value for Helm --set
    """
    # Convert to string and escape special characters
    escaped = str(v)
    escaped = escaped.replace('\\', '\\\\')  # Backslash must be first
    escaped = escaped.replace('{', '\\{')
    escaped = escaped.replace('}', '\\}')
    return escaped


def flatten_structure(data, parent_key=''):
    """
    Flatten a nested structure (dicts and lists) to dot notation with array indices.

    Examples:
        {"a": {"b": "c"}} -> {"a.b": "c"}
        {"a": [1, 2, 3]} -> {"a[0]": "1", "a[1]": "2", "a[2]": "3"}
        {"a": [{"b": "c"}, {"b": "d"}]} -> {"a[0].b": "c", "a[1].b": "d"}

    Args:
        data: Dictionary or list to flatten
        parent_key: Parent key prefix

    Returns:
        Flattened dictionary with dot-notation keys and properly escaped values
    """
    items = []

    if isinstance(data, dict):
        for k, v in data.items():
            # Build the new key with dot notation
            new_key = f"{parent_key}.{k}" if parent_key else k

            if isinstance(v, dict):
                # Recursively flatten nested dictionaries
                items.extend(flatten_structure(v, new_key).items())
            elif isinstance(v, list):
                # Recursively flatten lists with index notation
                items.extend(flatten_structure(v, new_key).items())
            elif v is None:
                # Skip null values
                continue
            else:
                # Leaf value - escape and add
                items.append((new_key, escape_helm_value(v)))

    elif isinstance(data, list):
        for idx, item in enumerate(data):
            # Build the new key with array index notation
            new_key = f"{parent_key}[{idx}]"

            if isinstance(item, dict):
                # Recursively flatten nested dictionaries
                items.extend(flatten_structure(item, new_key).items())
            elif isinstance(item, list):
                # Recursively flatten nested lists
                items.extend(flatten_structure(item, new_key).items())
            elif item is None:
                # Skip null values
                continue
            else:
                # Leaf value - escape and add
                items.append((new_key, escape_helm_value(item)))

    return dict(items)


def main():
    """Read JSON query from stdin, flatten the input, and output to stdout."""
    try:
        # Read query from stdin (Terraform external data source format)
        query = json.loads(sys.stdin.read())

        # The input data is in the "input" key as a JSON string
        input_json = query.get("input", "{}")
        data = json.loads(input_json)

        # Flatten the structure
        flattened = flatten_structure(data)

        # Terraform external data source requires all values to be strings
        # and the output must have a "flattened" key containing the result
        output = {
            "flattened": json.dumps(flattened)
        }

        # Output as JSON
        print(json.dumps(output))
        sys.exit(0)

    except json.JSONDecodeError as e:
        # Return error in expected format
        print(json.dumps({"error": f"Invalid JSON input: {e}"}), file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        # Return error in expected format
        print(json.dumps({"error": f"Error: {e}"}), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
```

**Comment by ppapapetrou76** [comment]
@nijdarshan can you please resolve the conflicts?

**Comment by nijdarshan** [comment]
hi @ppapapetrou76 

conflicts resolved, please check thanks!
